modules/contact/src/constraints/ComputeDynamicFrictionalForceLMMechanicalContact.C

```diff
@@ -0,0 +1,371 @@
+    const DofObject * const dof)
+{
+  // Get normal LM
+  const auto normal_dof_index = dof->dof_number(_sys.number(), _var->number(), 0);
+  const ADReal & weighted_gap = *_weighted_gap_ptr;
+  ADReal contact_pressure = (*_sys.currentSolution())(normal_dof_index);
+  Moose::derivInsert(contact_pressure.derivatives(), normal_dof_index, 1.);
+
+  // Get friction LMs
+  std::array<const ADReal *, 2> & tangential_vel = _tangential_vel_ptr;
+  std::array<dof_id_type, 2> friction_dof_indices;
+  std::array<ADReal, 2> friction_lm_values;
+
+  const unsigned int num_tangents = 2;
+  for (const auto i : make_range(num_tangents))
+  {
+    friction_dof_indices[i] = dof->dof_number(_sys.number(), _friction_vars[i]->number(), 0);
+    friction_lm_values[i] = (*_sys.currentSolution())(friction_dof_indices[i]);
+    Moose::derivInsert(friction_lm_values[i].derivatives(), friction_dof_indices[i], 1.);
+  }
+
+  // Get normalized c and c_t values (if normalization specified
+  const Real c = _normalize_c ? _c / *_normalization_ptr : _c;
+  const Real c_t = _normalize_c ? _c_t / *_normalization_ptr : _c_t;
+
+  ADReal dof_residual;
+  ADReal dof_residual_dir;
+
+  // Primal-dual active set strategy (PDASS)
+  if (contact_pressure < _epsilon)
+  {
+    dof_residual = friction_lm_values[0];
+    dof_residual_dir = friction_lm_values[1];
+  }
+  else
+  {
+    const Real epsilon_sqrt = 1.0e-14;
+
+    const auto lamdba_plus_cg = contact_pressure + c * weighted_gap;
+    std::array<ADReal, 2> lambda_t_plus_ctu;
+    lambda_t_plus_ctu[0] = friction_lm_values[0] + c_t * *tangential_vel[0] * _dt;
+    lambda_t_plus_ctu[1] = friction_lm_values[1] + c_t * *tangential_vel[1] * _dt;
+
+    const auto term_1_x =
+        std::max(_mu * lamdba_plus_cg,
+                 std::sqrt(lambda_t_plus_ctu[0] * lambda_t_plus_ctu[0] +
+                           lambda_t_plus_ctu[1] * lambda_t_plus_ctu[1] + epsilon_sqrt)) *
+        friction_lm_values[0];
+
+    const auto term_1_y =
+        std::max(_mu * lamdba_plus_cg,
+                 std::sqrt(lambda_t_plus_ctu[0] * lambda_t_plus_ctu[0] +
+                           lambda_t_plus_ctu[1] * lambda_t_plus_ctu[1] + epsilon_sqrt)) *
+        friction_lm_values[1];
+
+    const auto term_2_x = _mu * std::max(0.0, lamdba_plus_cg) * lambda_t_plus_ctu[0];
+
+    const auto term_2_y = _mu * std::max(0.0, lamdba_plus_cg) * lambda_t_plus_ctu[1];
+
+    dof_residual = term_1_x - term_2_x;
+    dof_residual_dir = term_1_y - term_2_y;
+  }
+
+  if (_subproblem.currentlyComputingJacobian())
+  {
+    _assembly.processDerivatives(dof_residual, friction_dof_indices[0], _matrix_tags);
+    _assembly.processDerivatives(dof_residual_dir, friction_dof_indices[1], _matrix_tags);
+  }
+  else
+  {
+    _assembly.cacheResidual(friction_dof_indices[0], dof_residual.value(), _vector_tags);
+    _assembly.cacheResidual(friction_dof_indices[1], dof_residual_dir.value(), _vector_tags);
```

Comment:
I have a theory that this could be one of the reasons why automatic scaling convergence has always been so bad for you. `cacheResidual` does not apply the variable's scaling factor but `processDerivatives` does. There is a new API in `Assembly` called `processResidual` and it has the same argument ordering as `processDerivatives` (but takes `_vector_tags` instead of `_matrix_tags`). Let's switch to using `processResidual` here and then let's make sure that we do this everywhere in this PR ... and also in the non-dynamic classes. This is something that has only recently come to my attention and I haven't gotten around to fixing up contact ... because I forgot I use this design here

Reply:
Is `processResidual` in `FVInterfaceKernel`?

Reply:
There is a `FVInterfaceKernel::processResidual` but you want to call `_assembly.processResidual`

Reply:
Along the lines of what you explain above: Are the corresponding displacement residuals properly scaled? (I can't see where this is happening)

Reply:
A class like this is only computing the LM residual (well at least that's typically what the 'LM' in the class name has indicated). The displacement residuals, unless `computeResidual` and `computeJacobian` have been overridden, are taken care of in `ADMortarConstraint` and this happens behind the scenes after `accumulateTaggedLocalResidual`

Reply:
The displacement residuals are also taken care of now in `ADMortarLagrangeConstraint`. But this class follows the same design as `ADMortarConstraint`, including calls to `accumulateTaggedLocalResidual`.

Did some initial testing and didn't notice significant improvement with `automatic_scaling` using `processResidual`, but will keep looking.

Reply:
It's totally possible that unscaled will still fare better since the residuals will be scaled more "physically" and automatic scaling, by default, goes wholly by the Jacobian.

At least these changes should make results of `-snes_test_jacobian` better when automatic scaling is present